documentation/general/torn-sdk.md

```diff
@@ -0,0 +1,109 @@
+# Torn .NET SDK
+
+## Terminology
+
+- msbuild: refers the .NET Framework based msbuild included with Visual Studio.
+- dotnet msbuild: refers to the .NET Core based msbuild included with the .NET SDK.
+- analyzers: this document will use analyzers to refer to both analyzers and generators.
+
+## Summary
+
+Visual Studio and .NET SDK are separate products but they are intertwined in command line and design time build scenarios as different components are loaded from each product.
+
+| Scenario | Loads Roslyn | Loads Analyzers / Generators |
+| --- | --- | --- |
+| msbuild | From Visual Studio | From .NET SDK |
+| dotnet msbuild | From .NET SDK | From .NET SDK |
+| Visual Studio Design Time | From Visual Studio | From Both |
+
+Generally this mixing of components is fine because Visual Studio will install a .NET SDK that is functionally paired with it. That is the compiler and analyzers are the same hence mixing poses no real issue. For example 17.10 installs .NET SDK 8.0.3xx, 17.9 installs .NET SDK 8.0.2xx, etc ... However when the .NET SDK is not paired with the Visual Studio version,then compatibility issues can, and will, arise. Particularly when the .NET SDK is _newer_ than Visual Studio customers will end up with the following style of error:
+
+> CSC : warning CS9057: The analyzer assembly '..\dotnet\sdk\8.0.200\Sdks\Microsoft.NET.Sdk.Razor\source-
+generators\Microsoft.CodeAnalysis.Razor.Compiler.SourceGenerators.dll' references version '4.9.0.0' of the compiler, which is newer
+than the currently running version '4.8.0.0'.
+
+To address this issue we are going to separate Visual Studio and the .NET SDK in build scenarios. That will change our matrix to the following:
+
+| Scenario | Loads Roslyn | Loads Analyzers / Generators |
+| --- | --- | --- |
+| msbuild | From .NET SDK | From .NET SDK |
+| dotnet msbuild | From .NET SDK | From .NET SDK |
+| Visual Studio Design Time | From Visual Studio | From Visual Studio |
+
+Specifically we will be
+
+1. Changing msbuild to use a compiler from the .NET SDK
+2. Changing Visual Studio to use analyzers from Visual Studio
+
+In addition to making our builds more reliable this will also massively simplify our analyzer development strategy. Analyzers following this model can always target the latest Roslyn version without the need for complicated multi-targeting.
+
+## Detailed Design
+
+### MSBuild using .NET SDK Compiler
+
+The .NET SDK will start producing a package named Microsoft.NETSdk.Compiler.Roslyn. This will contain a .NET Framework version of the Roslyn compiler that matches .NET Core version. This will be published as a part of the .NET SDK release process meaning it's available for all publicly available .NET SDKs.
```

Comment:
I did the bad thing and named something: Microsoft.NETSdk.Compiler.Roslyn. 

Given my history of naming this will not last and a new name will be chosen. I don't really have any objections to a better name but ... whomever picks the new name is responsible for fixing up all the occurrences in this document :smile:

Reply:
[Microsoft.NET.Sdk* packages](https://www.nuget.org/packages?q=microsoft.net.sdk) exist (with a dot between NET and Sdk unlike your proposed name) but they are used for dotnet workloads I think

Reply:
Is the package specific to framework? IE, it will _only_ have the framework compilers (arm and x86)?

Reply:
Initially it will only have x64 framework compilers. Those will still run just fine an ARM64 via emulation. If there is significant ask we can later start shipping an ARM64 variant but I'd rather wait for that need to arise.

Reply:
Going to go with @jjonescz suggestion on the naming.

Reply:
I wasn't suggesting a name, I just pointed out a similar name already exists. I'm not sure if we can reuse the same prefix for this package.

Reply:
May I suggest Microsoft.Net.Sdk.Compilers.Toolset. That matches other sdk naming of other packages as mentioned above and also matches the naming of the compiler toolset package we already have. I don't see Roslyn used commonly in our package names.

Reply:
That is fine by me but note that when we generalize Roslyn as Compiler there is often some pushback from people that Roslyn is not the only .NET language. That is why I put Roslyn in the name here. But I'm guessing we're the only compiler that realistically has this type of problem given how the SDK worsk today.